### PR TITLE
Add gp.pandas package in GreenplumPython

### PR DESCRIPTION
**Pull request by Quasimewdo**
This patch adds a pandas package to provide a pandas-compatible API, in which
the classes and functions has the same names and parameters as the equivalence in pandas.
all classes and functions act as proxies to the "native" API

### CODE REVIEW

File: greenplumpython/pandas/dataframe.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+
+class DataFrame:
+    def __init__(
+        self,
+        data: Union[
+            dataframe.DataFrame, List[Union[Tuple[Any, ...], Dict[str, Any]]], Dict[str, List[Any]]
+        ],
+        columns: Optional[List[str]] = None,
+        con: Optional[db.Database] = None,
```
Comment (yihong0618) [inline greenplumpython/pandas/dataframe.py:18]: conn maybe better name?

File: greenplumpython/pandas/dataframe.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        schema: Union[str, None] = None,
+        if_exists: Literal["fail", "replace", "append"] = "fail",
+        index: bool = False,  # Not Used
+        index_label: Union[str, List[str]] = None,  # Not Used
+        chunksize: Union[int, None] = None,  # Not Used
+        dtype: Union[Dict[str, type], None] = None,  # Not Used
+        method: Literal[None, "multi"] = None,  # Not Used  # type: ignore
+    ) -> int:
+        assert index is False, "DataFrame in GreenplumPython.pandas does not have an index column"
+        assert if_exists in ["fail", "replace"], "Only support if_exists = 'fail' or 'replace'"
```
Comment (yihong0618) [inline greenplumpython/pandas/dataframe.py:42]: if_exists: Literal["fail", "replace", "append"] = "fail"

assert if_exists in ["fail", "replace"], "Only support if_exists = 'fail' or 'replace'" 
we do not support append?

File: greenplumpython/pandas/dataframe.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        table_name = name if schema is None else (schema + "." + name)
+        with con.connect() as connexion:  # type: ignore
```
Comment (yihong0618) [inline greenplumpython/pandas/dataframe.py:44]: connexion -> connection

File: greenplumpython/pandas/dataframe.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    columns: Optional[list[str]] = None,
+    chunksize: Optional[int] = None,
+):
+    database = db.Database(url=str(con.url))
+    try:
+        database.execute(f"SELECT * FROM {sql}")
+        df = database.create_dataframe(table_name=sql)
+        return DataFrame(df, con=database)
+    except:
+        return DataFrame(dataframe.DataFrame(query=sql, db=database))
```
Comment (yihong0618) [inline greenplumpython/pandas/dataframe.py:172]: why try, except here?

File: tox.ini [hunk 5]
```diff
@@ -15,6 +15,7 @@ deps =
 [testenv:test]
 deps = {[base]deps}
 commands =
+    pip install sqlalchemy
```
Comment (yihong0618) [inline tox.ini:18]: why we need install sqlalchemy here since we have already put it in requirements?

Comment (Quasimewdo) [reply tox.ini:18]: Forgot to remove it, note it

File: greenplumpython/pandas/dataframe.py (hunk 4 above)
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:172]: To see if sql it is a table name or a query

File: greenplumpython/pandas/dataframe.py (hunk 3 above)
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:44]: Thx, a little confusion between french and english

File: greenplumpython/pandas/dataframe.py (hunk 1 above)
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:18]: Noted

File: greenplumpython/pandas/dataframe.py (hunk 2 above)
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:42]: We may support, will bring modifs

File: greenplumpython/pandas/dataframe.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+import greenplumpython.db as db
+import greenplumpython.group as groupby
+import greenplumpython.order as orderby
+
+
+class DataFrame:
+    def __init__(
+        self,
+        data: Union[
+            dataframe.DataFrame, List[Union[Tuple[Any, ...], Dict[str, Any]]], Dict[str, List[Any]]
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:25]: We should probably hide the constructor to user. But we will have probably an extra attribut `conn` or `db` in our `DataFrame.from_xxx()`

File: greenplumpython/pandas/dataframe.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+                self._dataframe = conn.create_dataframe(columns=data)
+            else:
+                self._dataframe = conn.create_dataframe(rows=data, column_names=columns)
+
+    def to_sql(
+        self,
+        name: str,
+        conn: engine,
+        schema: Union[str, None] = None,
+        if_exists: Literal["fail", "replace", "append"] = "fail",
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:44]: Noted

File: greenplumpython/pandas/dataframe.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+                    text(
+                        f"""
+                            INSERT INTO {table_name}
+                            {query}
+                        """
+                    )
+                )
+            else:
+                if if_exists == "replace":
+                    connection.execute(text(f"DROP TABLE IF EXISTS {table_name}"))
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:100]: Noted

File: greenplumpython/pandas/dataframe.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+        """
+        assert index is False, "DataFrame in GreenplumPython.pandas does not have an index column"
+        table_name = name if schema is None else (schema + "." + name)
+        with conn.connect() as connection:  # type: ignore
+            query = self._dataframe._build_full_query()  # type: ignore
+            if if_exists == "append":
+                result = connection.execute(  # type: ignore
+                    text(
+                        f"""
+                            INSERT INTO {table_name}
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:93]: Noted

File: greenplumpython/pandas/dataframe.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+import greenplumpython.dataframe as dataframe
+import greenplumpython.db as db
+import greenplumpython.order as orderby
+
+
+class DataFrame:
+    """Representation of GreenplumPython Pandas Compatible DataFrame object."""
+
+    @classmethod
+    def from_sql(cls, sql: str, conn: str):
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:21]: Yes we should, noted

File: greenplumpython/pandas/dataframe.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+        """
+        c = super().__new__(cls)
+        c._dataframe = df
+        return c
+
+    def __init__(self) -> None:
+        """
+        :meta private:
+        """
+        self._dataframe: dataframe.DataFrame = None
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/pandas/dataframe.py:40]: There will be type unknown error raised by lint

**Review by beeender — COMMENTED** [review]
Please enable the example test for the newly added `dataframe.py` in `tox.ini`
(docstyle check should be enabled already I think.)

**Review by beeender — APPROVED** [review]
LOOKS NICE!

**Pull request merged**